### PR TITLE
Improve Kafka serialization docs and error handling

### LINKED ISSUE #341: Error deserializing key/value for partition - Error deserializing Avro message for id
Hi 

 

I am using Akka-Stream-Kafka for consuming messages from Kafka in a Play framework application. Here are the details:

 

build.sbt

libraryDependencies += "com.typesafe.akka" %% "akka-stream-kafka" % "0.17"

 

And here is the setting for consumer

val consumerSettings = ConsumerSettings(system, new StringDeserializer, new KafkaAvroDeserializer())
  .withBootstrapServers("IP_ADDRESS:9092")
  .withProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,"org.apache.kafka.common.serialization.StringDeserializer")
  .withProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,"io.confluent.kafka.serializers.KafkaAvroDeserializer")
  .withProperty(KafkaAvroDeserializerConfig.SPECIFIC_AVRO_READER_CONFIG, "true")
  .withProperty("schema.registry.url", "http://IP_ADDRESS:8081")
  //.withProperty("auto.commit.enable", "false")
  .withGroupId("index.consumers")
  .withProperty("enable.auto.commit", "false")

 

While deserializing I get error. Here is the details:

 

[ERROR] [09/15/2017 19:50:13.610] [play-dev-mode-akka.kafka.default-dispatcher-12] [akka://play-dev-mode/system/kafka-consumer-1] Exception when polling from consumer

org.apache.kafka.common.errors.SerializationException: Error deserializing key/value for partition tst2-0 at offset 3

Caused by: org.apache.kafka.common.errors.SerializationException: Error deserializing Avro message for id 141

Caused by: java.lang.NullPointerException

         at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:122)

         at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:93)

         at io.confluent.kafka.serializers.KafkaAvroDeserializer.deserialize(KafkaAvroDeserializer.java:55)

         at org.apache.kafka.common.serialization.ExtendedDeserializer$Wrapper.deserialize(ExtendedDeserializer.java:65)

         at org.apache.kafka.common.serialization.ExtendedDeserializer$Wrapper.deserialize(ExtendedDeserializer.java:55)

         at org.apache.kafka.clients.consumer.internals.Fetcher.parseRecord(Fetcher.java:918)

         at org.apache.kafka.clients.consumer.internals.Fetcher.access$2600(Fetcher.java:93)

         at org.apache.kafka.clients.consumer.internals.Fetcher$PartitionRecords.fetchRecords(Fetcher.java:1095)

         at org.apache.kafka.clients.consumer.internals.Fetcher$PartitionRecords.access$1200(Fetcher.java:944)

         at org.apache.kafka.clients.consumer.internals.Fetcher.fetchRecords(Fetcher.java:567)

         at org.apache.kafka.clients.consumer.internals.Fetcher.fetchedRecords(Fetcher.java:528)

         at org.apache.kafka.clients.consumer.KafkaConsumer.pollOnce(KafkaConsumer.java:1110)

         at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1043)

         at akka.kafka.KafkaConsumerActor.tryPoll$1(KafkaConsumerActor.scala:257)

         at akka.kafka.KafkaConsumerActor.poll(KafkaConsumerActor.scala:303)

         at akka.kafka.KafkaConsumerActor.akka$kafka$KafkaConsumerActor$$receivePoll(KafkaConsumerActor.scala:232)

         at akka.kafka.KafkaConsumerActor$$anonfun$receive$1.applyOrElse(KafkaConsumerActor.scala:151)

         at akka.actor.Actor.aroundReceive(Actor.scala:514)

         at akka.actor.Actor.aroundReceive$(Actor.scala:512)

         at akka.kafka.KafkaConsumerActor.aroundReceive(KafkaConsumerActor.scala:73)

         at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)

         at akka.actor.ActorCell.invoke(ActorCell.scala:496)

         at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)

         at akka.dispatch.Mailbox.run(Mailbox.scala:224)

         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

         at java.lang.Thread.run(Thread.java:748)

 

[error] p.c.s.c.WebSocketFlowHandler - WebSocket flow threw exception

java.lang.Exception: Consumer actor terminated

         at akka.kafka.internal.SingleSourceLogic.$anonfun$preStart$1(SingleSourceLogic.scala:54)

         at akka.kafka.internal.SingleSourceLogic.$anonfun$preStart$1$adapted(SingleSourceLogic.scala:40)

         at akka.stream.stage.GraphStageLogic$StageActor.internalReceive(GraphStage.scala:185)

         at akka.stream.stage.GraphStageLogic$StageActor.$anonfun$callback$1(GraphStage.scala:152)

         at akka.stream.stage.GraphStageLogic$StageActor.$anonfun$callback$1$adapted(GraphStage.scala:152)

         at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:443)

         at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:453)

         at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:546)

         at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:725)

         at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:740)

 

BTW, there are some warnings in the Console while debugging:

 

[info] play.api.Play - Application started (Dev)

[warn] o.a.k.c.c.ConsumerConfig - The configuration 'schema.registry.url' was supplied but isn't a known config.

[warn] o.a.k.c.c.ConsumerConfig - The configuration 'specific.avro.reader' was supplied but isn't a known config.

[warn] o.a.k.c.c.ConsumerConfig - The configuration 'auto.commit.enable' was supplied but isn't a known config.

 

Can someone please help?

### LINKED ISSUE #356: Serialization exception not surfaced in "Consumer actor terminated" exception
We ran into a situation in which a topic contained a message containing an invalid schema id.

The Kafka Avro deserializer complained as follows:

Caused by: org.apache.kafka.common.errors.SerializationException: Error retrieving Avro schema for id 4
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Schema not found; error code: 40403

Rather than resulting in something we could handle with Supervision (we'd like to skip over this element), instead the consumer actor was terminated, and the exception is very generic and contains no cause.

In the onFailure handler of the materialized Future[Done], we saw:

Caused by: java.lang.Exception: Consumer actor terminated
at akka.kafka.internal.SubSourceLogic$SubSourceStage$$anon$1$$anonfun$preStart$2.apply(SubSourceLogic.scala:150) 

At that point there is no way to skip the element by resuming, and the stream is stuck because it is stopped (and restarted by our host actor) with no offset committed. The "Consumer actor terminated" exception has no cause value.

Obviously we are working to prevent such invalid messages, but would like to be able to handle them with supervision. Also posted to akka-user.
https://groups.google.com/forum/#!topic/akka-user/WP-CSoyuIaA

### PR DESCRIPTION
**Pull request by ennru**
* Add a new documentation page on **Serialization** showing how to use the Schema Registry (to solve #341, #342, #558)
* Avro Serialization exercised in Scala tests and Java tests
* Pass Serialization errors from the `KafkaConsumerActor` to the stages involved in the request (fixes #356), other stages on a shared actor will survive

Fixes #341, #342, #356, and #558

### CODE REVIEW

File: docs/src/main/paradox/serialization.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Serialization
+
+The general recommendation for de-/serialization of messages is to use byte arrays as value and do the de-/serialization in a `map` operation in the Akka Stream instead of implementing it directly in Kafka de-/serializers.
+
+
+## Avro with Schema Registry
+
+If you want to use [Confluent's Schema Registry](https://docs.confluent.io/current/schema-registry/docs/index.html), you need to include the dependency on `kafka-avro-serializer` as shown below. It is not available from Maven Central, that's why Confluent's repository has to be specified. These examples use `kafka-avro-seriazlizer` version $confluent.version$.
```
Comment (2m) [inline docs/src/main/paradox/serialization.md]: As this paragraph follows the general recommendation to use serialization in `map` operators, the implicit understanding is that this paragraph is about using avro inside `map` operator.

We should add a contrasting paragraph to the general recommendation that it is possible to specify (de)serializers to in Producer/Consumer settings and here how that can be used with Avro.

File: tests/src/test/scala/docs/scaladsl/SerializationSpec.scala [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    case 0 => key
+    case 1 => name
+    case index => throw new AvroRuntimeException(s"Unknown index: $index")
+  }
+
+  override def put(i: Int, v: scala.Any): Unit = i match {
+    case 0 => key = asString(v)
+    case 1 => name = asString(v)
+    case index => throw new AvroRuntimeException(s"Unknown index: $index")
+  }
```
Comment (2m) [inline tests/src/test/scala/docs/scaladsl/SerializationSpec.scala]: Ooh. Mutable serialization.

File: tests/src/test/scala/docs/scaladsl/SerializationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  private def asString(v: Any) =
+    v match {
+      case utf8: Utf8 => utf8.toString
+      case _ => v.asInstanceOf[String]
+    }
+  override def getSchema: Schema = SampleAvroClass.SCHEMA$
+}
+
+object SampleAvroClass {
+  val SCHEMA$ : Schema =
```
Comment (2m) [inline tests/src/test/scala/docs/scaladsl/SerializationSpec.scala]: Why the dollar?

Comment (ennru) [reply tests/src/test/scala/docs/scaladsl/SerializationSpec.scala]: That's the tradition in generated Avro classes.

File: tests/src/test/java/docs/javadsl/SerializationTest.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  public static void beforeClass() {
+    // Schema registry uses Glassfish which uses java.util.logging
+    SLF4JBridgeHandler.removeHandlersForRootLogger();
+    SLF4JBridgeHandler.install();
+    startEmbeddedKafka(kafkaPort, zooKeeperPort, schemaRegistryPort, 1);
+  }
+
+  @Test
+  public void jacksonDeSer() throws Exception {
+    final String topic = createTopic(nextNumber(), 1, 1);
```
Comment (2m) [inline tests/src/test/java/docs/javadsl/SerializationTest.java]: This now can use `1` again as the last number in the topic is incremented using the same AtomicInteger across all Java tests.

File: tests/src/test/java/docs/javadsl/SerializationTest.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+    List<SampleData> result =
+        control.drainAndShutdown(ec).toCompletableFuture().get(1, TimeUnit.SECONDS);
+    assertThat(result, is(samples));
+    assertThat(result.size(), is(samples.size()));
+  }
+
+  @Test
+  public void avroDeSerMustWorkWithSchemaRegistry() throws Exception {
+    final String topic = createTopic(nextNumber(), 1, 1);
```
Comment (2m) [inline tests/src/test/java/docs/javadsl/SerializationTest.java]: Same here, this can be changed back to `1`.

**Pull request merged**